src/frequenz/sdk/microgrid/_power_managing/_shifting_matryoshka.py

```diff
@@ -0,0 +1,303 @@
+            return Power.zero()
+
+
+class ShiftingMatryoshka(BaseAlgorithm):
+    """The ShiftingMatryoshka algorithm."""
+
+    def __init__(self, max_proposal_age: timedelta) -> None:
+        """Create a new instance of the matryoshka algorithm."""
+        self._max_proposal_age_sec = max_proposal_age.total_seconds()
+        self._component_buckets: dict[frozenset[int], set[Proposal]] = {}
+        self._target_power: dict[frozenset[int], Power] = {}
+
+    def _calc_targets(
+        self,
+        proposals: set[Proposal],
+        system_bounds: SystemBounds,
+        priority: int | None = None,
+    ) -> tuple[Power | None, Bounds[Power]]:
+        """Calculate the target power and bounds for the given components.
+
+        Args:
+            proposals: The proposals for the given components.
+            system_bounds: The system bounds for the components in the proposal.
+            priority: The priority of the actor for which the target power is calculated.
+
+        Returns:
+            The new target power and bounds for the components.
+        """
+        lower_bound = (
+            system_bounds.inclusion_bounds.lower
+            if system_bounds.inclusion_bounds
+            # if a target power exists from a previous proposal, and the system bounds
+            # have become unavailable, force the target power to be zero, by narrowing
+            # the bounds to zero.
+            else Power.zero()
+        )
+        upper_bound = (
+            system_bounds.inclusion_bounds.upper
+            if system_bounds.inclusion_bounds
+            else Power.zero()
+        )
+
+        if not proposals:
+            return None, Bounds[Power](lower=lower_bound, upper=upper_bound)
+
+        available_bounds = Bounds[Power](lower=lower_bound, upper=upper_bound)
+        top_pri_bounds: Bounds[Power] | None = None
+
+        target_power = Power.zero()
+        for next_proposal in sorted(proposals, reverse=True):
+            if priority is not None and next_proposal.priority <= priority:
+                break
+
+            if upper_bound < lower_bound:
+                break
+
+            proposal_lower = next_proposal.bounds.lower or lower_bound
+            proposal_upper = next_proposal.bounds.upper or upper_bound
+            proposal_power = next_proposal.preferred_power
+
+            if proposal_upper < proposal_lower:
+                continue
+
+            if proposal_power and (
+                proposal_power < proposal_lower or proposal_power > proposal_upper
+            ):
+                continue
+
+            if proposal_lower >= upper_bound:
+                proposal_lower = upper_bound
+                proposal_upper = upper_bound
+            elif proposal_upper <= lower_bound:
+                proposal_lower = lower_bound
+                proposal_upper = lower_bound
+
+            lower_bound = max(lower_bound, proposal_lower)
+            upper_bound = min(upper_bound, proposal_upper)
```

Comment:
Are all these checks for `proposal_power`, `proposal_lower` and `proposal_upper` intentionally covering for an eventual `0.0` besides `None` or was the intention to only match `None`? Since in the code below, you are actually checking for `None` explicitly, I guess it is intentional, but maybe it would be best to add some comment to say it explicitly (either way).

Reply:
> intentionally covering for an eventual `0.0` besides `None` or was the intention to only match `None`

I don't understand what that means.  Of course, I can add comments for these, though.

Reply:
added more comments, also fixed a bug.

Reply:
I was talking specifically about using a nullable int as a boolean implicitly, the `next_proposal.bounds.lower or lower_bound`,  `f proposal_power and ...`, etc. In these cases both `0` and `None` will be considered `False`, so the question was if this was considered. Sometimes I feel it might be good to have a linting rule that requires writing `if (proposal_power == 0 or proposal_power is None) and (...` (but only sometimes, because other times I guess it can be overkill :laughing: )

Reply:
Oh damn, you're right, we can't `continue` here if `proposal_power == 0`.  Stupid Python and its weird implicit booleans.  I will make a PR next week.

> Sometimes I feel it might be good to have a linting rule that requires writing `if (proposal_power == 0 or proposal_power is None) and (...` (but only sometimes, because other times I guess it can be overkill 😆 )

That sounds like a great idea. It is not overkill. Expressing our intentions clearly is always the way to go.

Reply:
Upvote so it is ready when we switch to ruff :)

* https://github.com/astral-sh/ruff/issues/10949

Reply:
Looks like I had already changed this before merging: https://github.com/frequenz-floss/frequenz-sdk-python/blob/6d106d149aaa3c686370d997fa4793f601806b0e/src/frequenz/sdk/microgrid/_power_managing/_shifting_matryoshka.py#L136

So nothing to do.